src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -942,3 +944,213 @@ fixCopyCommand(Archive *fout, PQExpBuffer copyBuf, TableInfo *tbinfo, bool isFro
+			 * the sqlvariant column. These extra columns are carefully handled
+			 * so that they do not interfere with expected dump behaviour.
+			*/
+			if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+			{
+					appendPQExpBuffer(buf, ", sys.SQL_VARIANT_PROPERTY(%s, 'BaseType')", tbinfo->attnames[i]);
+					appendPQExpBuffer(buf, ", sys.datalength(%s)", fmtId(tbinfo->attnames[i]));
+					is_sqlvariant_column[nfields_new] = true;
+					nfields_new = nfields_new + 2;
+			}
+		}
+		nfields_new++;
+	}
+	*(nfields) = nfields_new;
+	return;
+}
+
+/*
+ * fixBbfSqlvariantData:
+ * Modify the values in INSERT query that dump sql_variant column table data so
+ * that metadata is preserved during restore. If default value is not used or 
+ * the value is not NULL, we add a cast expression to typecast the value to its
+ * appropriate basetype.
+ */
+void
+fixBbfSqlvariantData(Archive *fout, PGresult* res, PQExpBuffer q, char* attgenerated, int tuple, int *orig_field, int *field)
+{
+	if (*field > 0)
+		archputs(", ", fout);
+	if (attgenerated[*orig_field])
+	{
+		archputs("DEFAULT", fout);
+		(*orig_field)++;
+		*field = *(field) + 2;
+		return;
+	}
+	if (PQgetisnull(res, tuple, *field))
+	{
+		archputs("NULL", fout);
+		(*orig_field)++;
+		*field = *(field) + 2;
+		return;
+	}
+	resetPQExpBuffer(q);
+	appendStringLiteralAH(q,
+				PQgetvalue(res, tuple, *field),
+								fout);
+	castSqlvariantToBasetype(res, fout, q, tuple, *field);
+
+	*orig_field = *(orig_field) + 1;
+	*field = *(field) + 2;
+	return;
+}
+
+/*
+ * hasSqlvariantColumn:
+ * Returns true if any of the columns in table is a sqlvariant data type column
+ */
+
+bool
+hasSqlvariantColumn(TableInfo *tbinfo)
+{
+	for(int i=0; i<tbinfo->numatts; i++)
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+					return true;
+	return false;
+}
+
+/*
+ * getMbstrlen:
+ * returns the length of a multibyte string
+ */
+static int
+getMbstrlen(const char* mbstr, Archive* fout)
+{
+	int len = 0;
+	if(!mbstr)
+		return 0;
+	while (*mbstr){
+		mbstr += PQmblen(mbstr, fout->encoding);
+		len++;
+	}
+	return len;
+}
+
+/*
+ * castSqlvariantToBasetype:
+ * Modify INSERT query in dump file by adding a CAST expression for a sql_variant
+ * column data entry in order to preserve the metadata of data type otherwise
+ * lost during restore.
+ */
+static void
+castSqlvariantToBasetype(PGresult* res, Archive* fout, PQExpBuffer q, int tuple, int field)
+{
+	char* value = PQgetvalue(res, tuple, field);
+	char* type = PQgetvalue(res, tuple, field+1);
+	int datalength = atoi(PQgetvalue(res, tuple, field+2));
+	int precision;
+	int scale;
+	int i;
+
+	archprintf(fout, "CAST(%s AS ", q->data);
+
+	/* data types defined in sys schema should be handled separately */
+	if(!pg_strcasecmp(type, "datetime") || !pg_strcasecmp(type, "datetimeoffset")
```

Comment:
Nit: commenting here but applies to everywhere, please add a space between `if/else if` and `(`.